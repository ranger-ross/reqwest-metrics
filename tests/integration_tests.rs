use metrics_util::debugging::{DebuggingRecorder, Snapshotter};
use reqwest_metrics::MetricsMiddleware;
use reqwest_middleware::{ClientBuilder, reqwest};
use tokio::test;
use wiremock::{
    Mock, MockServer, ResponseTemplate,
    matchers::{method, path},
};

const SNAPSHOT_FILTERS: [(&'static str, &'static str); 3] = [
    (
        r"Histogram\(\s*[\s\S]*?\s*\)",
        "Histogram([HISTOGRAM_VALUE])",
    ),
    (
        r#"Label\(\s*"port"\s*,\s*[\s\S]*?\s*\)"#,
        r#"Label("port", [PORT])"#,
    ),
    (r#"hash: \d*"#, "hash: [HASH]"),
];

#[test]
async fn basic() {
    let snapshotter = install_debug_recorder();

    let client = ClientBuilder::new(reqwest::Client::new())
        .with(MetricsMiddleware::new())
        .build();

    let mock_server = MockServer::start().await;

    Mock::given(method("GET"))
        .and(path("/hello"))
        .respond_with(ResponseTemplate::new(200))
        .mount(&mock_server)
        .await;

    let url = mock_server.uri();

    let res = client.get(format!("{url}/hello")).send().await.unwrap();
    assert_eq!(200, res.status().as_u16());

    let snapshot = snapshotter.snapshot();
    insta::with_settings!({filters => SNAPSHOT_FILTERS}, {
        insta::assert_debug_snapshot!(snapshot);
    });
}

fn install_debug_recorder() -> Snapshotter {
    let recorder = DebuggingRecorder::new();
    let snapshotter = recorder.snapshotter();
    recorder.install().unwrap();
    snapshotter
}
