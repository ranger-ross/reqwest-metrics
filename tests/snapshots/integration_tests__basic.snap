---
source: tests/integration_tests.rs
expression: snapshot
---
Snapshot(
    [
        (
            CompositeKey(
                Histogram,
                Key {
                    name: KeyName(
                        "http.client.request.duration",
                    ),
                    labels: [
                        Label(
                            "http.request.method",
                            "GET",
                        ),
                        Label(
                            "url.scheme",
                            "http",
                        ),
                        Label(
                            "network.protocol.name",
                            "http",
                        ),
                        Label(
                            "server.address",
                            "127.0.0.1",
                        ),
                        Label("server.port", [PORT]),
                        Label(
                            "network.protocol.version",
                            "1.1",
                        ),
                        Label(
                            "http.response.status_code",
                            "200",
                        ),
                    ],
                    hashed: true,
                    hash: [HASH],
                },
            ),
            Some(
                Seconds,
            ),
            Some(
                "Duration of HTTP client requests.",
            ),
            Histogram([HISTOGRAM_VALUE]),
        ),
    ],
)
